Load Run Evaluator

hinthornw opened the pull request:
Current problems:
1. Evaluating LLMs or Chat models isn't smooth. Even specifying 'generations' as the output inserts a redundant list into the eval template
2. Configuring input / prediction / reference keys in the `get_qa_evaluator` function is confusing. Unless you are using a chain with the default keys, you have to specify all the variables and need to reason about whether the key corresponds to the traced run's inputs, outputs or the examples inputs or outputs.


Proposal:
- Configure the run evaluator according to a model. Use the model type and input/output keys to assert compatibility where possible. Only need to specify a reference_key for certain evaluators (which is less confusing than specifying input keys)


When does this work:
- If you have your langchain model available (assumed always for run_on_dataset flow)
- If you are evaluating an LLM, Chat model, or chain
- If the LLM or chat models are traced by langchain (wouldn't work if you add an incompatible schema via the REST API)

When would this fail:
- Currently if you directly create an example from an LLM run, the outputs are generations with all the extra metadata present. A simple `example_key` and dumping all to the template could make the evaluations unreliable
- Doesn't help if you're not using the low level API
- If you want to instantiate the evaluator without instantiating your chain or LLM (maybe common for monitoring, for instance) -> could also load from run or run type though

What's ugly:
- Personally think it's better to load evaluators one by one since passing a config down is pretty confusing.
- Lots of testing needs to be added
- Inconsistent in that it makes a separate run and example input mapper instead of the original `RunEvaluatorInputMapper`, which maps a run and example to a single input. 

Example usage running the for an LLM, Chat Model, and Agent.

```
# Test running for the string evaluators
evaluator_names = ["qa", "criteria"]

model = ChatOpenAI()
configured_evaluators = load_run_evaluators_for_model(evaluator_names, model=model, reference_key="answer")
run_on_dataset(ds_name, model, run_evaluators=configured_evaluators)
```


<details>
  <summary>Full code with dataset upload</summary>
```
## Create dataset
from langchain.evaluation.run_evaluators.loading import load_run_evaluators_for_model
from langchain.evaluation import load_dataset
import pandas as pd

lcds = load_dataset("llm-math")
df = pd.DataFrame(lcds)

from uuid import uuid4
from langsmith import Client
client = Client()
ds_name = "llm-math - " + str(uuid4())[0:8]
ds = client.upload_dataframe(df, name=ds_name, input_keys=["question"], output_keys=["answer"])



## Define the models we'll test over
from langchain.llms import OpenAI
from langchain.chat_models import ChatOpenAI
from langchain.agents import initialize_agent, AgentType

from langchain.tools import tool

llm = OpenAI(temperature=0)
chat_model = ChatOpenAI(temperature=0)

@tool
    def sum(a: float, b: float) -> float:
        """Add two numbers"""
        return a + b
    
def construct_agent():
    return initialize_agent(
        llm=chat_model,
        tools=[sum],
        agent=AgentType.OPENAI_MULTI_FUNCTIONS,
    )

agent = construct_agent()

# Test running for the string evaluators
evaluator_names = ["qa", "criteria"]

models = [llm, chat_model, agent]
run_evaluators = []
for model in models:
    run_evaluators.append(load_run_evaluators_for_model(evaluator_names, model=model, reference_key="answer"))
    

# Run on LLM, Chat Model, and Agent
from langchain.client.runner_utils import run_on_dataset

to_test = [llm, chat_model, construct_agent]

for model, configured_evaluators in zip(to_test, run_evaluators):
    run_on_dataset(ds_name, model, run_evaluators=configured_evaluators, verbose=True)
```
</details>

hinthornw committed:
Add evaluator loader

hinthornw committed:
docs

hinthornw committed:
Add Better Errors for Comparison Chain

hinthornw committed:
Add Better Errors for Comparison Chain

hinthornw committed:
Merge

hinthornw committed:
Merge

hinthornw committed:
Log errors

hinthornw committed:
Merge branch 'wfh/re-add-errors' into wfh/to_run_evaluator

hinthornw committed:
Accept no 'reasoning' response in qa evaluator

hinthornw committed:
Merge branch 'wfh/re-add-qa-reasoning-handling' into wfh/to_run_evaluator

hinthornw committed:
Update Loader

hinthornw committed:
Merge branch 'wfh/load_evals' into wfh/to_run_evaluator

agola11 commented on langchain/evaluation/run_evaluators/string_run_evaluator.py:
```diff
@@ -0,0 +1,349 @@
+        return ["prediction", "input"]
+
+    @abstractmethod
+    def map(self, run: Run) -> Dict[str, Any]:
```
these should all be Dict[str, str]

hinthornw replied:
Not really for things like trajectory evals

hinthornw committed:
merge

agola11 commented on langchain/evaluation/run_evaluators/string_run_evaluator.py:
```diff
@@ -0,0 +1,349 @@
+            input_ = "\n\n".join(inputs["prompts"])
+        elif "prompt" in inputs:
+            input_ = inputs["prompt"]
+        elif "messages" in inputs:
```
I actually think this is a `List[List[dict]]`

agola11 replied:
but would be good to double check

agola11 replied:
Actually, if we are dealing with Runs downloaded from LangSmith, bc of the flatmap logic in the server, these would be List[dict], but otherwise if they are straight from the tracer, they can be List[List[dict]] so we probs need to handle both cases

hinthornw replied:
Yeah the message and generation formats are unreliable still

hinthornw committed:
add run eval

hinthornw committed:
Unset Cache in Tests

This is impacting other unit tests

hinthornw committed:
Merge branch 'wfh/unset_cache_in_tests' into wfh/to_run_evaluator

hinthornw committed:
add int tests

hinthornw committed:
merge

hinthornw committed:
add some updates from nuno

nfcampos committed:
Small updates to run_evaluator after adding dataset.data_type (#7279)

<!-- Thank you for contributing to LangChain!

Replace this comment with:
  - Description: a description of the change, 
  - Issue: the issue # it fixes (if applicable),
  - Dependencies: any dependencies required for this change,
- Tag maintainer: for a quicker response, tag the relevant maintainer
(see below),
- Twitter handle: we announce bigger features on Twitter. If your PR
gets announced and you'd like a mention, we'll gladly shout you out!

If you're adding a new integration, please include:
1. a test for the integration, preferably unit tests that do not rely on
network access,
  2. an example notebook showing its use.

Maintainer responsibilities:
  - General / Misc / if you don't know who to tag: @baskaryan
  - DataLoaders / VectorStores / Retrievers: @rlancemartin, @eyurtsev
  - Models / Prompts: @hwchase17, @baskaryan
  - Memory: @hwchase17
  - Agents / Tools / Toolkits: @hinthornw
  - Tracing / Callbacks: @agola11
  - Async: @agola11

If no one reviews your PR within a few days, feel free to @-mention the
same people again.

See contribution guidelines for more information on how to write/run
tests, lint, etc:
https://github.com/hwchase17/langchain/blob/master/.github/CONTRIBUTING.md
 -->

---------

Co-authored-by: William Fu-Hinthorn <13333726+hinthornw@users.noreply.github.com>
